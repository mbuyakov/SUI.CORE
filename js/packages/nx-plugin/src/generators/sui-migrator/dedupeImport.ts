import {tsquery} from "./tsqeury";
import {factory, Identifier, ImportDeclaration, ImportSpecifier, NamedImports, StringLiteral} from "typescript";
import {logRemap, printNode} from "./util";

type ImportData = {
  count: number,
  importSpecifiers: ImportSpecifier[],
  defaultImport?: Identifier
};

export function dedupeImport(content: string): string {
  const imports = tsquery.query<ImportDeclaration>(content, "ImportDeclaration");

  const importMap: { [index: string]: ImportData } = {};

  imports.forEach(importDeclaration => {
    const moduleName = (importDeclaration.moduleSpecifier as StringLiteral).text;
    const mapEntry: ImportData = (importMap[moduleName] = importMap[moduleName] || {count: 0, importSpecifiers: []});
    mapEntry.count++;
    mapEntry.defaultImport = importDeclaration.importClause.name || mapEntry.defaultImport;
    mapEntry.importSpecifiers.push(...(importDeclaration.importClause.namedBindings as NamedImports).elements);
  });

  Object.keys(importMap).forEach(moduleName => {
    if (importMap[moduleName].count > 1) {
      logRemap("dedupeImport", `Dedupe import from ${moduleName}`);
      const mapEntry = importMap[moduleName];

      // Leave only first import
      let isFirst = true;
      content = tsquery.remove(content, `ImportDeclaration:has(StringLiteral[value="${moduleName}"])`, () => {
        if (isFirst) {
          isFirst = false;
          return false;
        } else {
          return true;
        }
      }, true);

      content = tsquery.replace(content, `ImportDeclaration:has(StringLiteral[value="${moduleName}"])`, (node: ImportDeclaration) => {
        node = factory.updateImportDeclaration(
          node,
          undefined,
          factory.updateImportClause(
            node.importClause,
            node.importClause.isTypeOnly,
            mapEntry.defaultImport,
            factory.createNamedImports(mapEntry.importSpecifiers)
          ),
          node.moduleSpecifier,
          node.assertClause
        );

        return printNode(node)
          .replace("{ ", "{")
          .replace(" }", "}");
      });
    }
  });

  return content;
}
