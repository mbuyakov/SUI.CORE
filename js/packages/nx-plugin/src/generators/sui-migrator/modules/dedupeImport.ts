import {
  Identifier,
  ImportDeclaration,
  ImportSpecifier,
  isNamespaceImport,
  NamedImports,
  NamespaceImport,
  StringLiteral
} from "typescript";
import {logWithPrefix} from "../../../utils/logger";
// eslint-disable-next-line @nx/enforce-module-boundaries
import {
  astQuery,
  astRemove,
  astReplace,
  ParsedImportDeclaration,
  parseImportSpecifier,
  printImportDeclaration,
  printNodes
} from "@sui/lib-typescript-ast";

type ImportData = {
  count: number,
  importSpecifiers: ImportSpecifier[],
  nsImport?: NamespaceImport,
  defaultImport?: Identifier
};

export function dedupeImport(content: string): string {
  const imports = astQuery<ImportDeclaration>(content, "ImportDeclaration:has(ImportClause)");

  const importMap: { [index: string]: ImportData } = {};

  imports.forEach(importDeclaration => {
    const moduleName = (importDeclaration.moduleSpecifier as StringLiteral).text;
    const mapEntry: ImportData = (importMap[moduleName] = importMap[moduleName] || {count: 0, importSpecifiers: []});
    mapEntry.count++;
    mapEntry.defaultImport = importDeclaration?.importClause?.name || mapEntry.defaultImport;
    if (importDeclaration?.importClause?.namedBindings) {
      if (isNamespaceImport(importDeclaration.importClause.namedBindings)) {
        mapEntry.nsImport = importDeclaration.importClause.namedBindings;
      } else {
        mapEntry.importSpecifiers.push(...(importDeclaration.importClause.namedBindings as NamedImports).elements);
      }
    }
  });

  Object.keys(importMap).forEach(moduleName => {
    const mapEntry = importMap[moduleName];
    if (
      importMap[moduleName].count > 2 ||
      // If exactly two import - it might me smth like
      // import * as React from "react"; -- nsImport
      // import {useMemo} from "react";  -- importSpecifiers
      (importMap[moduleName].count == 2 && !(mapEntry.nsImport && (mapEntry.defaultImport || mapEntry.importSpecifiers)))
    ) {
      logWithPrefix("dedupeImport", `Dedupe import from ${moduleName}`);

      const mapEntry = importMap[moduleName];

      // Leave only first import
      let isFirst = true;
      content = astRemove(content, `ImportDeclaration:has(ImportClause):has(StringLiteral[value="${moduleName}"])`, () => {
        if (isFirst) {
          isFirst = false;
          return false;
        } else {
          return true;
        }
      }, true);

      content = astReplace(content, `ImportDeclaration:has(ImportClause):has(StringLiteral[value="${moduleName}"])`, () => {
        const rows: ParsedImportDeclaration[] = [];

        rows.push({
          import: mapEntry.defaultImport?.text,
          namedImports: mapEntry.importSpecifiers.map(it => parseImportSpecifier(it)),
          from: moduleName
        });

        if (mapEntry.nsImport) {
          rows.push({
            nsImport: mapEntry.nsImport.name.text,
            namedImports: [],
            from: moduleName
          });
        }

        return printNodes(rows.map(it => printImportDeclaration(it)));
      });
    }
  });

  return content;
}
