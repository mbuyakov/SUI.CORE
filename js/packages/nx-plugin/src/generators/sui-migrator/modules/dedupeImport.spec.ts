import {dedupeImport} from "./dedupeImport";

describe("dedupe import", () => {

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should leave as is", () => {
    const consoleLog = jest.spyOn(console, "log");
    const source = `
import {a} from "a";
import {b as b} from "b";
import c from "c";
import * as d from "d";
import "a";
    `.trim();
    expect(dedupeImport(source)).toBe(source);
    expect(consoleLog).not.toBeCalled();
  });

  it("should leave as in named import & namespace", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(dedupeImport(`
import {a} from "a";
import {b as b} from "b";
import c from "c";
import * as d from "a";
import "a";
    `.trim())).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(0);
  });

  it("should merge multiple named import", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(dedupeImport(`
import {a} from "a";
import {b as b} from "a";
import c from "c";
import * as d from "d";
import "a";
    `.trim())).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(1);
  });

  it("should merge named import & default", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(dedupeImport(`
import {a} from "a";
import {b as b} from "b";
import c from "a";
import * as d from "d";
import "a";
    `.trim())).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(1);
  });

  it("should merge named import & default + namespace on next line", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(dedupeImport(`
import {a} from "a";
import {b as b} from "a";
import {f} from "b";
import c from "a";
import * as d from "a";
import "a";
    `.trim())).toMatchSnapshot();
    expect(consoleLog).toBeCalledTimes(1);
  });

  it("should merge two same import", () => {
    const consoleLog = jest.spyOn(console, "log");
    expect(dedupeImport(`
import React from "react";
import React from "react";
    `).trim()).toBe(`
import React from "react";
    `.trim());
    expect(consoleLog).toBeCalledTimes(1);
  });
});
