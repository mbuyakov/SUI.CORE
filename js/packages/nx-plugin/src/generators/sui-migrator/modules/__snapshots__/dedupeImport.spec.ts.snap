// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dedupe import should leave as in named import & namespace 1`] = `
"import {a} from \\"a\\";
import {b as b} from \\"b\\";
import c from \\"c\\";
import * as d from \\"a\\";
import \\"a\\";"
`;

exports[`dedupe import should merge multiple named import 1`] = `
"import {a, b as b} from \\"a\\";
import c from \\"c\\";
import * as d from \\"d\\";
import \\"a\\";"
`;

exports[`dedupe import should merge named import & default + namespace on next line 1`] = `
"import c, {a, b as b} from \\"a\\";
import * as d from \\"a\\";
import {f} from \\"b\\";
import \\"a\\";"
`;

exports[`dedupe import should merge named import & default 1`] = `
"import c, {a} from \\"a\\";
import {b as b} from \\"b\\";
import * as d from \\"d\\";
import \\"a\\";"
`;
